spring-cloud-data-module-deployers/spring-cloud-data-module-deployer-yarn/src/test/java/org/springframework/cloud/data/module/deployer/yarn/YarnCloudAppStateMachineTests.java

```diff
@@ -0,0 +1,357 @@
+				.setHeader(YarnCloudAppStateMachine.HEADER_CLUSTER_ID, "fakeClusterId")
+				.setHeader(YarnCloudAppStateMachine.HEADER_COUNT, 1)
+				.setHeader(YarnCloudAppStateMachine.HEADER_MODULE, "fakeModule")
+				.setHeader(YarnCloudAppStateMachine.HEADER_DEFINITION_PARAMETERS, new HashMap<Object, Object>())
+				.build();
+
+		StateMachineTestPlan<States, Events> plan =
+				StateMachineTestPlanBuilder.<States, Events>builder()
+					.defaultAwaitTime(2)
+					.stateMachine(stateMachine)
+					.step()
+						.expectStateMachineStarted(1)
+						.expectStates(States.READY)
+						.and()
+					.step()
+						.sendEvent(message)
+						.expectStateChanged(7)
+						.expectStates(States.READY)
+						.and()
+					.build();
+		plan.test();
+
+		assertThat(yarnCloudAppService.getApplicationsLatch.await(2, TimeUnit.SECONDS), is(true));
+		assertThat(yarnCloudAppService.getApplicationsCount, is(1));
+
+		assertThat(yarnCloudAppService.getInstancesLatch.await(2, TimeUnit.SECONDS), is(true));
+		assertThat(yarnCloudAppService.getInstancesCount, is(2));
+
+		assertThat(yarnCloudAppService.pushApplicationLatch.await(2, TimeUnit.SECONDS), is(false));
+		assertThat(yarnCloudAppService.pushApplicationCount.size(), is(0));
+
+		assertThat(yarnCloudAppService.submitApplicationLatch.await(2, TimeUnit.SECONDS), is(true));
+		assertThat(yarnCloudAppService.submitApplicationCount.size(), is(1));
+		assertThat(yarnCloudAppService.submitApplicationCount.get(0).appVersion, is("fakeApp"));
+
+		assertThat(yarnCloudAppService.createClusterLatch.await(2, TimeUnit.SECONDS), is(true));
+		assertThat(yarnCloudAppService.createClusterCount.size(), is(1));
+		assertThat(yarnCloudAppService.createClusterCount.get(0).yarnApplicationId, is("fakeApplicationId"));
+		assertThat(yarnCloudAppService.createClusterCount.get(0).clusterId, is("fakeClusterId"));
+		assertThat(yarnCloudAppService.createClusterCount.get(0).count, is(1));
+		assertThat(yarnCloudAppService.createClusterCount.get(0).module, is("fakeModule"));
+		assertThat(yarnCloudAppService.createClusterCount.get(0).definitionParameters.size(), is(0));
+
+		assertThat(yarnCloudAppService.startClusterLatch.await(2, TimeUnit.SECONDS), is(true));
+		assertThat(yarnCloudAppService.startClusterCount.size(), is(1));
+		assertThat(yarnCloudAppService.startClusterCount.get(0).yarnApplicationId, is("fakeApplicationId"));
+		assertThat(yarnCloudAppService.startClusterCount.get(0).clusterId, is("fakeClusterId"));
+
+		context.close();
+	}
+
+	@Configuration
+	static class Config {
+
+		@Bean
+		TaskExecutor taskExecutor() {
+			ThreadPoolTaskExecutor taskExecutor = new ThreadPoolTaskExecutor();
+			taskExecutor.setCorePoolSize(1);
+			return taskExecutor;
+		}
+
+	}
+
+	private static class TestYarnCloudAppService implements YarnCloudAppService {
+
+		String app = null;
+		String instance = null;
+
+		volatile CountDownLatch getApplicationsLatch = new CountDownLatch(1);
```

Comment:
I don't see these being mutated anywhere - would `final` be more appropriate here?